src/boot/lib/ast.ml

```diff
@@ -113,6 +113,14 @@ and const =
   | Cgensym
   | Ceqsym of Symb.t option
   | Csym2hash
+  (* Map intrinsics *)
+  (* NOTE(Linnea, 2021-01-27): Obj.t denotes the type of the internal map (I was so far unable to express it properly) *)
+  | CMap of (tm -> tm -> int) * Obj.t
```

Comment:
I think this is a pretty reasonable way to do it, and probably good enough for our use cases at the moment, and possibly even more efficient than what I discuss below (though that's hard to know without benchmarking of course).

Given how Map works in OCaml we essentially need to hide the type of the map, but not necessarily the type of the value. My first thought on this would be with an existential variable, essentially making the constructor something like this:

```ocaml
| CMap : (module Map.S with t = 't and key = tm) -> tm t' -> tm
(* or maybe something like *)
| CMap : (type t) -> (module Map.S with t = t and key = tm) -> tm t -> tm
```

(I don't actually know how much of that is valid OCaml, in particular I'm not sure if type variables can have any other kind than `*`. )

With that we could pattern match out the module and the map with appropriate locally abstract types:

```ocaml
match tm with
| CMap (type t, module M : Map.S with t = t and key = tm, m) ->
  CMap (module M, M.insert someKey someValue m)
```

----------------------

At this point, if we want to have a safe interface, it's probably more reasonable to just wrap the `Obj.t` stuff in such an interface, or make a new implementation, or see if `batteries` has a reasonable map we could use.

Reply:
I did try some similar to what you suggested, as an attempt to store the module in a variable as a first-class module, instead of always creating a new module using `Map.Make` as in the current implementation. What I got stuck on was expressing that the key type is `tm`, that is, essentially the "`module Map.S with key = tm`" part in what you wrote. What I tried was declaring a module type that is like a `Map.S` with key type `tm`:
```
module type MapModTm = sig
  include Map.S
  type key = tm
end
```
but this gives an error saying `Illegal shadowing of included type ... The value of_seq has no valid type if key/1254 is shadowed`.

However, there might be some other way to do it.